### DISCUSSION TITLE
System.Diagnostics.ActivityListener vs Processor/Sampler

### DISCUSSION THREAD

**Discussion by danelson** [open-telemetry/opentelemetry-dotnet#3912] [General]
It seems that [System.Diagnostics.ActivityListener](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activitylistener?view=net-7.0) provides similar functionality to that of a custom Processor/Sampler.

- Should one be favored over another? Are there exceptions?
- Is this just a matter of which dependencies one wishes to take?

**Comment by cijothomas** [comment]
OTel .NET SDK (shipped in this repo) is built on top of ActivityListener, following the OTel specification. Nothing preventing one to write own listener as long as they understand the use cases and implications.

**Reply by danelson** [reply]
Let me be more explicit. I believe the following code samples result in similar (identical?) behavior . Should one approach be preferred over the other?

**OpenTelemetry**
```
public class BaggageProcessor : BaseProcessor<Activity>
{
    public override void OnStop(Activity activity)
    {
        foreach (var item in Baggage.Current)
        {
            activity.SetTag(item.Key, item.Value);
        }
    }
}
```

**System.Diagnostics.DiagnosticSource**
```
ActivityListener baggageListener = new ActivityListener()
{
    ShouldListenTo = _ => true,
    ActivityStopped = activity =>
    {
        foreach (var item in Baggage.Current)
        {
            activity.SetTag(item.Key, item.Value);
        }
    },
    Sample = (ref ActivityCreationOptions<ActivityContext> _) => ActivitySamplingResult.AllData
};
ActivitySource.AddActivityListener(baggageListener);
```

**Reply by cijothomas** [reply]
The activity listener you have does not run samplers. Not sure what is the end goal and why are you writing own listeners? If you describe the end goal, we can help.

**Reply by danelson** [reply]
In this instance it is simply setting attributes (tags) based on what exists in the baggage which I believe both of those samples accomplish. I dislike that there appears to be 2 ways to do this. Generally I haven't considered many other use cases.

I believe one advantage to `ActivityListener` is that it can be "registered" at any time, unlike the processor which needs to be added when the provider is configured.

Perhaps I can follow an approach of always use a custom Sampler/Processor unless I need this kind of "delayed registration".

**Reply by cijothomas** [reply]
This repo is the OTel repo, so our answer is always to use OpenTelemetry SDK. For API, it is recommended to use the DiagnosticSource API directly (https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Api#introduction-to-opentelemetry-net-tracing-api)

I do not think I can give an advice about using ActivityListener vs OTel.

**Reply by cijothomas** [reply]
_unlike the processor which needs to be added when the provider is configured._

https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/customizing-the-sdk#processors--exporters Processors can be added to the provider any time, even after it was built.

**Reply by danelson** [reply]
Thank you, that was a fundamental misunderstanding on my part. I think then I can just generally prefer the SDK.

I appreciate all the responses.
